luarules/gadgets/unit_projectile_leash.lua

```diff
@@ -0,0 +1,181 @@
+function gadget:GetInfo()
+	return {
+		name = "Projectile Over-Range and Leashing",
+		desc = "Destroys projectiles if they exceed defined ranges",
+		author = "SethDGamre",
+		layer = 1,
+		enabled = true
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+---- unit customParams ----
+-- use weaponDef customparams.projectile_overrange_distance to destroy projectiles exceeding its limit.
+-- use weaponDef customparams.projectile_leash_range to destroy projectiles when they exceed range/projectile_overrange_distance and projectile_leash_range relative to unit position.
+
+---- optional customParams ----
+-- use weaponDef customparams.weaponDef.customParams.projectile_destruction_method = "string" to change how projectiles are destroyed.
+-- "explode" (default if undefined) detonates the projectile.
+-- "descend" moves the projectile downward until it is destroyed by collision event.
+
+--static values
+local forcedDescentUpdateFrames = math.ceil(Game.gameSpeed / 5)
+local compoundingMultiplier = 1.1 --compounding multiplier that influences the arc at which projectiles are forced to descend
+local descentSpeedStartingMultiplier = 0.15
+local flightTimeSlop = 0.9
+local projectileWatchModulus = math.ceil(Game.gameSpeed / 3)
+
+--functions
+local spGetUnitPosition = Spring.GetUnitPosition
+local mathRandom = math.random
+local mathCeil = math.ceil
+local spGetProjectilePosition = Spring.GetProjectilePosition
+local spSetProjectileCollision = Spring.SetProjectileCollision
+local spGetProjectileVelocity = Spring.GetProjectileVelocity
+local spSetProjectileVelocity = Spring.SetProjectileVelocity
+
+--tables
+local defWatchTable = {}
+local projectileMetaData = {}
+local flightTimeWatch = {}
+local projectileWatch = {}
+local forcedDescentTable = {}
+local killQueue = {}
+
+--variables
+local gameFrame = 0
+
+
+for weaponDefID, weaponDef in pairs(WeaponDefs) do
+	if weaponDef.customParams.projectile_leash_range or weaponDef.customParams.projectile_overrange_distance then
```

Comment:
getting `local customParams = weaponDef.customParams` here before the first "if" will help a lot with readability